server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -139,14 +147,41 @@ public NamedAnalyzer getNamedAnalyzer(String analyzerName) {
         return this.indexAnalyzers.get(analyzerName);
     }
 
+    //TODO This is only used in tests, we may want to look into replacing those usages?
     public DocumentMapperParser documentMapperParser() {
         return this.documentMapperParser;
     }
 
+    public Mapper.TypeParser.ParserContext parserContext() {
+        return parserContextSupplier.get();
+    }
+
     DocumentParser documentParser() {
         return this.documentParser;
     }
 
+    Map<Class<? extends MetadataFieldMapper>, MetadataFieldMapper> getMetadataMappers() {
+        Map<Class<? extends MetadataFieldMapper>, MetadataFieldMapper> metadataMappers = new LinkedHashMap<>();
```

Comment:
I think this can be simplified as follows:
1. if the existing mapper is null, iterate through mapperRegistry.getMetadataMapperParsers and call getDefault on each one.
2. if not, just return the metadata mappers from the existing mapper

The mixing and matching is unnecessary, because the index created version is fixed between the new mapper and the previous mapper, and so the set of metadata mappers for the two will be the same.